include/matx/core/iterator.h

```diff
@@ -0,0 +1,230 @@
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "matx/core/defines.h"
+
+namespace matx {
+/**
+ * @brief Iterator around operators for libraries that can take iterators as input (CUB).
+ * 
+ * @tparam T Data type
+ * @tparam RANK Rank of tensor
+ * @tparam Desc Descriptor for tensor
+ * 
+ */
+template <typename OperatorType>
+struct RandomOperatorIterator {
+  using self_type = RandomOperatorIterator<OperatorType>;
+  using value_type = typename OperatorType::scalar_type;
+  using scalar_type = typename OperatorType::scalar_type;
+  // using stride_type = std::conditional_t<is_tensor_view_v<OperatorType>, typename OperatorType::desc_type::stride_type,
+  //                         index_t>;
+  using stride_type = index_t;
+  using pointer = value_type*;
+  using reference = value_type&;
+  using iterator_category = std::random_access_iterator_tag;
+  using difference_type = typename std::iterator<iterator_category, value_type>::difference_type;
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ RandomOperatorIterator(const OperatorType &t) : t_(t), offset_(0) { }
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ RandomOperatorIterator(const OperatorType &t, stride_type offset) : t_(t), offset_(offset) {}
+
+  /**
+   * @brief Dereference value at a pre-computed offset
+   * 
+   * @return Value at offset 
+   */
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ value_type operator*() const
+  {
+    if constexpr (OperatorType::Rank() == 0) {
+      return t_.operator()();
+    }
+    else {
+      auto arrs = detail::GetIdxFromAbs(t_, offset_);
+      return detail::mapply([&](auto &&...args) {
+          return t_.operator()(args...);
+        }, arrs);     
+    }
+  }  
+
+
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type operator+(difference_type offset) const
+  {
+    return self_type{t_, offset_ + offset};
+  }
+
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ value_type operator[](difference_type offset) const
+  {
+    return *self_type{t_, offset_ + offset};
+  }  
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__  self_type operator++(int)
```

Comment:
is this correct?  wouldn't ++(int) increment by int?

Reply:
Not sure what you mean, but this is incrementing by integer on offset_, but it returns the old value (iterator)

Reply:
oh right i had forgot what operator++(int) was.  it's pre vs post ++.